src/agents/MonarchAgent.sol

```diff
@@ -27,4 +47,57 @@ contract MonarchAgentV1 is IMonarchAgent {
 
         emit RebalancerSet(msg.sender, rebalancer);
     }
+
+    /**
+     * @notice Users can revoke the authorization of a rebalancer
+     */
+    function revoke() external {
+        delete rebalancers[msg.sender];
+
+        emit RebalancerSet(msg.sender, address(0));
+    }
+
+    /**
+     * @notice Rebalances the user's position from one market to another
+     * @param onBehalf The user to rebalance the position for
+     * @param token The token to rebalance
+     * @param fromMarkets The markets to withdraw assets from
+     * @param toMarkets The markets to supply assets to
+     */
+    function rebalance(
+        address onBehalf,
+        address token,
+        RebalanceMarketParams[] calldata fromMarkets,
+        RebalanceMarketParams[] calldata toMarkets
+    ) external payable onlyRebalancer(onBehalf) {
+        require(onBehalf == msg.sender || morphoBlue.isAuthorized(onBehalf, address(this)), ErrorsLib.AGENT_NOT_AUTHORIZED);
+        require(fromMarkets.length > 0, ErrorsLib.ZERO_MARKET);
+        require(toMarkets.length > 0, ErrorsLib.ZERO_MARKET);
+
+        int256 tokenDelta;
+
+        for (uint256 i; i < fromMarkets.length; ++i) {
+            require(fromMarkets[i].market.loanToken == token, ErrorsLib.INVALID_TOKEN);
+            (uint256 assetsWithdrawn,) =
+                morphoBlue.withdraw(fromMarkets[i].market, fromMarkets[i].assets, 0, onBehalf, address(this));
+            tokenDelta += int256(assetsWithdrawn);
+        }
+
+        for (uint256 i; i < toMarkets.length; ++i) {
+            require(toMarkets[i].market.loanToken == token, ErrorsLib.INVALID_TOKEN);
+            _approveMaxTo(toMarkets[i].market.loanToken, address(morphoBlue));
```

Comment:
this line should be move outside of the loop

Reply:
done

Reply:
testing comment here